table.go

```diff
@@ -243,27 +243,25 @@ func (rt *RoutingTable) NearestPeers(id ID, count int) []peer.ID {
 	// It's assumed that this also protects the buckets.
 	rt.tabLock.RLock()
 
-	// Get bucket at cpl index or last bucket
-	var bucket *Bucket
+	// Get bucket index or last bucket
 	if cpl >= len(rt.Buckets) {
 		cpl = len(rt.Buckets) - 1
 	}
-	bucket = rt.Buckets[cpl]
 
 	pds := peerDistanceSorter{
 		peers:  make([]peerDistance, 0, 3*rt.bucketsize),
 		target: id,
 	}
-	pds.appendPeersFromList(bucket.list)
-	if pds.Len() < count {
-		// In the case of an unusual split, one bucket may be short or empty.
-		// if this happens, search both surrounding buckets for nearby peers
-		if cpl > 0 {
-			pds.appendPeersFromList(rt.Buckets[cpl-1].list)
-		}
-		if cpl < len(rt.Buckets)-1 {
-			pds.appendPeersFromList(rt.Buckets[cpl+1].list)
-		}
+
+	// Add peers from the target bucket and _less_ specific buckets until we
+	// have enough peers.
+	for i := cpl; i < len(rt.Buckets) && pds.Len() < count; i++ {
```

Comment:
**For my own understanding**

We add buckets with higher CPL("less specific") first  & then the buckets with a lower CPL("more specific") so that we maximize the number of peers closer to the target, right ?

Reply:
It's the opposite. More specific to less specific (i.e. shorter distance, to higher distance).

Reply:
I think @aarshkshah1992 may be right. Isn't the logic reversed?

* higher `i` == closer.
* higher `cpl` == closer.

We should start with `rt.Bucket[cpl]` (closest to target) and move further?

Reply:
So. Funny thing. My logic was wrong but the code is correct.

1. We calculate how many bits the target key shares with us.
2. We find the bucket for peers with that many shared bits.
3. Buckets to the _right_ have more shared bits with us but the same number of shared bits with the target.
4. Buckets to the _left_ have fewer shared bits with us _and_ the target.

Therefore, we want to go right first.

Reply:
This mimics the behavior of a catch-all bucket. We should probably be _shrinking_ buckets somewhere but I don't think we are.

Reply:
Ok, so the correct logic is too slow. I'm going to be _technically_ incorrect and call it out in the code.